src/compiler/compiler.rs

```diff
@@ -847,11 +847,11 @@ where
             if output.status.success() {
                 if let Ok(stdout) = String::from_utf8(output.stdout) {
                     if stdout.starts_with("rustc ") {
-                        return Some(stdout);
+                        return Some(Ok(stdout));
                     }
                 }
             }
-            None
+            Some(Err(String::from_utf8_lossy(&output.stderr).into_owned()))
```

Comment:
FYI there's an existing [`run_input_output`](https://github.com/mozilla/sccache/blob/6ba6f6c15c106768b914a7697a763e2232fa253a/src/util.rs#L170) function that might be useful for you here, it returns a specific error when the process exits with a failure status that contains stdout/stderr.

Reply:
Good to know! It doesn't quite map to what we're doing here, though.

Reply:
It seems better to adopt `run_input_output` to our needs here, perhaps by setting up the core `command` object with `env_clear().envs(...).args(...)` and calling into `run_input_output` rather than reproducing the code in a separate place.  Dealing with external commands is tricky enough that we want to centralize that logic in a single place.

Reply:
We can use `run_input_output`, but the main advantage seems to be providing the process' output if it fails in a relevant error type that we hopefully wouldn't have to mess with ourselves. In this case though we're trying to fail when the process fails, but also if we're given something that looks like rustc but doesn't produce output like it's rustc, so we'd have to inspect and convert the output of `run_input_output` anyway, which is why I mentioned above it doesn't quite map to what we're doing here. The duplication was already around and doesn't seem egregious, but at this point I guess I'll just write that version up and see how it actually looks.